tests/Tests.Grpc/CustomAutoDataAttributes.cs

```diff
@@ -0,0 +1,39 @@
+using System;
+using AutoFixture;
+using AutoFixture.AutoMoq;
+using AutoFixture.NUnit3;
+
+namespace Tests
+{
+    [AttributeUsage(AttributeTargets.Method)]
+    public class CustomAutoDataAttribute : AutoDataAttribute
+    {
+        public CustomAutoDataAttribute() : base(FixtureHelpers.CreateFixture)
+        {
+        }
+    }
+
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
+    public class CustomInlineAutoDataAttribute : InlineAutoDataAttribute
+    {
+        public CustomInlineAutoDataAttribute(params object[] args) : base(FixtureHelpers.CreateFixture, args)
+        {
+        }
+    }
+
+    internal static class FixtureHelpers
```

Comment:
Seeing the word "helper" triggers me. 🙂 It usually means the code is "homeless".
This might be over-kill but something I've grown fond of lately is using the C# type system.
Here's how I would write the same bit.

```cs
internal class FixtureFactory : CompositeCustomization
{
    private readonly Func<IFixture> CreateFixture;

    public FixtureFactory(params ICustomization[] customizations)
        : this(() => new Fixture(), customizations)
    {
    }

    public FixtureFactory(Func<IFixture> factory, params ICustomization[] customizations)
        : base(customizations)
    {
        this.CreateFixture = factory;
    }

    public static implicit operator Func<IFixture>(FixtureFactory source)
    {
        return () => source.CreateFixture().Customize(source);
    }
}
```
and then to crate an implicit implementation
```cs
internal class ImplicitFixtureFactory : FixtureFactory
{
    public ImplicitFixtureFactory()
        : base(new AutoMoqCustomization())
    {
    }
}
```
and then you can use it like
```cs
[AttributeUsage(AttributeTargets.Method)]
public class CustomAutoDataAttribute : AutoDataAttribute
{
    public CustomAutoDataAttribute() : base(new ImplicitFixtureFactory())
    {
    }
}
```

Reply:
It looks very promising but ho do you handle additional customizations like Register, Inject, Customize and so on?

I'm thinking of cases like this one: https://github.com/insight-architectures/dotnet-utils/blob/main/tests/Tests.ServiceModel.Proxies/CustomAutoDataAttributes.cs#L28-L50

Reply:
You can pass in as many `ICustomization` instances as you like. 
```cs
internal class ImplicitFixtureFactory : FixtureFactory
{
    public ImplicitFixtureFactory()
        : base(
            new AutoMoqCustomization
            {
                ConfigureMembers = true,
                GenerateDelegates = true
            }, new DomainCustomization())
    {
    }
}

internal class DomainCustomization : ICustomization
{
   public void Customize(IFixture fixture)
   {
      fixture.Register((Uri uri) => new EndpointAddress(uri.ToString()));
      fixture.Register<Binding>(() => new BasicHttpBinding());
      fixture.Register<IServiceCollection>(() => new ServiceCollection());
      fixture.Register((EndpointAddress address, Binding binding) => new Mock<ChannelFactory<ITestService>>(binding, address));
   }
}
```
If you have specific customizations you can chain it with the fixture factory instance.

```cs
[AttributeUsage(AttributeTargets.Method)]
public class CustomAutoDataAttribute : AutoDataAttribute
{
    public CustomAutoDataAttribute() 
        : base(new ImplicitFixtureFactory().Cusotmize(new MyOtherCustomization()))
    {
    }
}
```

Reply:
Oh, I see the point now :) Interesting take :)